src/swappy_allocator.rs

```diff
@@ -0,0 +1,1195 @@
+{
+    budget: usize,
+    path: PathBuf,
+
+    mem_usage: AtomicUsize,
+    mem_allocator: A,
+
+    maybe_swapped: AtomicBool,
+    page_seq: AtomicU32,
+    pages: Mutex<Vec<Page>>,
+}
+
+/// An [`Allocator`] implementation that has a memory budget and can be swapped
+/// out.
+///
+/// The allocations of its internal metadata are not managed (i.e. allocated via
+/// `std::alloc::Global`). Do NOT use it as the global allocator.
+#[derive(Clone)]
+pub struct SwappyAllocator<A: Allocator>(Arc<SwappyAllocatorCore<A>>);
+
+impl<A: Allocator> SwappyAllocator<A> {
+    pub fn new_over(path: &Path, budget: usize, alloc: A) -> SwappyAllocator<A> {
+        if path.exists() {
+            if let Err(e) = std::fs::remove_dir_all(&path) {
+                error!(
+                    "Failed to clean up old swap directory: {}. \
+                    There might be obsolete swap files left in {}.",
+                    e,
+                    path.display()
+                );
+            }
+        }
+        let core = SwappyAllocatorCore {
+            budget,
+            path: path.into(),
+            mem_usage: AtomicUsize::new(0),
+            mem_allocator: alloc,
+            maybe_swapped: AtomicBool::new(false),
+            page_seq: AtomicU32::new(0),
+            pages: Mutex::new(Vec::new()),
+        };
+        SwappyAllocator(Arc::new(core))
+    }
+
+    #[inline]
+    pub fn memory_usage(&self) -> usize {
+        self.0.mem_usage.load(Ordering::Relaxed)
+    }
+
+    #[inline]
+    fn is_swapped(&self, ptr: NonNull<u8>, exaustive_check: bool) -> bool {
+        // Ordering: `maybe_swapped` must be read after the pointer is available.
+        std::sync::atomic::fence(Ordering::Acquire);
+        self.0.maybe_swapped.load(Ordering::Relaxed)
+            && (!exaustive_check || self.0.pages.lock().iter().any(|page| page.contains(ptr)))
+    }
+
+    #[inline]
+    fn allocate_swapped(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+        let mut pages = self.0.pages.lock();
+        if pages.is_empty() {
+            self.0.maybe_swapped.store(true, Ordering::Relaxed);
+            // Ordering: `maybe_swapped` must be set before the page is created.
+            std::sync::atomic::fence(Ordering::Release);
+            pages.push(
+                Page::new(
+                    &self.0.path,
+                    self.0.page_seq.fetch_add(1, Ordering::Relaxed),
+                    std::cmp::max(DEFAULT_PAGE_SIZE, layout.size()),
+                )
+                .ok_or(AllocError)?,
+            );
+        }
+        debug_assert!(self.0.maybe_swapped.load(Ordering::Relaxed));
+        match pages.last_mut().unwrap().allocate(layout) {
+            None => {
+                pages.push(
+                    Page::new(
+                        &self.0.path,
+                        self.0.page_seq.fetch_add(1, Ordering::Relaxed),
+                        std::cmp::max(DEFAULT_PAGE_SIZE, layout.size()),
```

Comment:
A larger page will be created. And you may worry that the page will still not be large enough because of alignment padding. But I checked that the address returned from mmap will have the maximum alignment, i.e. it can always fulfil the request.